Sources/NIOCore/AsyncChannel/AsyncChannelInboundStreamChannelHandler.swift

```diff
@@ -86,12 +135,50 @@ internal final class NIOAsyncChannelInboundStreamChannelHandler<InboundIn: Senda
 
     @inlinable
     func channelRead(context: ChannelHandlerContext, data: NIOAny) {
-        self.buffer.append(self.unwrapInboundIn(data))
+        let unwrapped = self.unwrapInboundIn(data)
+
+        switch self.transformation {
+        case .sync(let transformation):
+            do {
+                try self.buffer.append(transformation(unwrapped))
+            } catch {
+                context.close(promise: nil)
+                context.fireErrorCaught(error)
+                return
+            }
+        case .protocolNegotiation(let protocolNegotiation):
+            let unsafeSelf = UnsafeTransfer(self)
+            let unsafeContext = UnsafeTransfer(context)
+            protocolNegotiation(unwrapped)
+                .whenComplete { result in
```

Comment:
Can you add a comment why the unsafe transfers are okay here? At the moment it relies on the negotiation result being on the correct EL.